File: server/src/test/java/com/larsreimann/api_editor/transformation/UnusedAnnotationProcessorTest.java
package com.larsreimann.api_editor.transformation;

import com.larsreimann.api_editor.model.AnnotatedPythonClass;
import com.larsreimann.api_editor.model.AnnotatedPythonFunction;
import com.larsreimann.api_editor.model.AnnotatedPythonModule;
import com.larsreimann.api_editor.model.AnnotatedPythonPackage;
import com.larsreimann.api_editor.model.UnusedAnnotation;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import static com.larsreimann.api_editor.util.PackageDataFactoriesKt.createAnnotatedPythonClass;
import static com.larsreimann.api_editor.util.PackageDataFactoriesKt.createAnnotatedPythonFunction;
import static com.larsreimann.api_editor.util.PackageDataFactoriesKt.createAnnotatedPythonModule;
import static com.larsreimann.api_editor.util.PackageDataFactoriesKt.createAnnotatedPythonPackage;

class UnusedAnnotationProcessorTest {
    @Test
    void shouldRemoveAnnotatedClass() {
        // given
        UnusedAnnotationProcessor unusedAnnotationProcessor =
            new UnusedAnnotationProcessor();

        AnnotatedPythonClass annotatedTestClass =
            createAnnotatedPythonClass("annotatedTestClass");
        annotatedTestClass.getAnnotations().add(UnusedAnnotation.INSTANCE);

        AnnotatedPythonClass testClass =
            createAnnotatedPythonClass("testClass");

        AnnotatedPythonModule testModule =
            createAnnotatedPythonModule("testModule");
        testModule.getClasses().add(testClass);

        AnnotatedPythonPackage testPackage =
            createAnnotatedPythonPackage("testPackage");
        testPackage.getModules().add(testModule);

        // when
        testPackage.accept(unusedAnnotationProcessor);
        AnnotatedPythonPackage modifiedPackage =
            unusedAnnotationProcessor.getModifiedPackage();

        // then
        Assertions.assertTrue(
            modifiedPackage
                .getModules().get(0)
                .getClasses()
                .size() == 1
                &&
                modifiedPackage
                    .getModules().get(0)
                    .getClasses().get(0)
                    .getName()
                    .equals("testClass")
        );
    }

    @Test
    void shouldRemoveAnnotatedFunction() {
        // given
        UnusedAnnotationProcessor unusedAnnotationProcessor =
            new UnusedAnnotationProcessor();

        AnnotatedPythonFunction annotatedTestFunction =
            createAnnotatedPythonFunction("annotatedTestFunction");
        annotatedTestFunction.getAnnotations().add(UnusedAnnotation.INSTANCE);

        AnnotatedPythonFunction testFunction =
            createAnnotatedPythonFunction("testFunction");

        AnnotatedPythonModule testModule =
            createAnnotatedPythonModule("testModule");
        testModule.getFunctions().add(testFunction);
        testModule.getFunctions().add(annotatedTestFunction);

        AnnotatedPythonPackage testPackage =
            createAnnotatedPythonPackage("testPackage");
        testPackage.getModules().add(testModule);

        // when
        testPackage.accept(unusedAnnotationProcessor);
        AnnotatedPythonPackage modifiedPackage =
            unusedAnnotationProcessor.getModifiedPackage();

        // then
        Assertions.assertTrue(
            modifiedPackage
                .getModules().get(0)
                .getFunctions()
                .size() == 1
                &&
                modifiedPackage.getModules().get(0)
                    .getFunctions().get(0)
                    .getName()
                    .equals("testFunction")
        );
    }

    @Test
    void shouldRemoveMethod() {
        // given
        UnusedAnnotationProcessor unusedAnnotationProcessor =
            new UnusedAnnotationProcessor();

        AnnotatedPythonFunction annotatedTestMethod =
            createAnnotatedPythonFunction("annotatedTestMethod");
        annotatedTestMethod.getAnnotations().add(UnusedAnnotation.INSTANCE);

        AnnotatedPythonFunction testMethod =
            createAnnotatedPythonFunction("testMethod");

        AnnotatedPythonClass testClass =
            createAnnotatedPythonClass("testClass");
        testClass.getMethods().add(annotatedTestMethod);
        testClass.getMethods().add(testMethod);

        AnnotatedPythonModule testModule =
            createAnnotatedPythonModule("testModule");
        testModule.getClasses().add(testClass);

        AnnotatedPythonPackage testPackage =
            createAnnotatedPythonPackage("testPackage");
        testPackage.getModules().add(testModule);

        // when
        testPackage.accept(unusedAnnotationProcessor);
        AnnotatedPythonPackage modifiedPackage =
            unusedAnnotationProcessor.getModifiedPackage();

        // then
        Assertions.assertTrue(
            modifiedPackage
                .getModules().get(0)
                .getClasses().get(0)
                .getMethods()
                .size() == 1
                &&
                modifiedPackage
                    .getModules().get(0)
                    .getClasses().get(0)
                    .getMethods().get(0)
                    .getName()
                    .equals("testMethod")
        );
    }
}

File: server/src/test/kotlin/com/larsreimann/api_editor/transformation/UnusedAnnotationProcessorTest.kt
package com.larsreimann.api_editor.transformation

import com.larsreimann.api_editor.model.UnusedAnnotation
import com.larsreimann.api_editor.mutable_model.PythonClass
import com.larsreimann.api_editor.mutable_model.PythonFunction
import com.larsreimann.api_editor.mutable_model.PythonModule
import com.larsreimann.api_editor.mutable_model.PythonPackage
import io.kotest.matchers.collections.shouldBeEmpty
import io.kotest.matchers.collections.shouldHaveSize
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

class UnusedAnnotationProcessorTest {
    private lateinit var testMethod: PythonFunction
    private lateinit var testGlobalFunction: PythonFunction
    private lateinit var testClass: PythonClass
    private lateinit var testPackage: PythonPackage

    @BeforeEach
    fun reset() {
        testMethod = PythonFunction(name = "testMethod")
        testGlobalFunction = PythonFunction(name = "testGlobalFunction")
        testClass = PythonClass(
            name = "TestClass",
            methods = listOf(testMethod)
        )
        testPackage = PythonPackage(
            distribution = "testPackage",
            name = "testPackage",
            version = "1.0.0",
            modules = listOf(
                PythonModule(
                    name = "testModule",
                    classes = listOf(testClass),
                    functions = listOf(testGlobalFunction)
                )
            )
        )
    }

    @Test
    fun `should process UnusedAnnotations of classes`() {
        testClass.annotations += UnusedAnnotation

        testPackage.processUnusedAnnotations()

        val modules = testPackage.modules
        modules.shouldHaveSize(1)

        modules[0].classes.shouldBeEmpty()
    }

    @Test
    fun `should process UnusedAnnotations of global functions`() {
        testGlobalFunction.annotations += UnusedAnnotation

        testPackage.processUnusedAnnotations()

        val modules = testPackage.modules
        modules.shouldHaveSize(1)

        modules[0].functions.shouldBeEmpty()
    }

    @Test
    fun `should process UnusedAnnotations of methods`() {
        testMethod.annotations += UnusedAnnotation

        testPackage.processUnusedAnnotations()

        val modules = testPackage.modules
        modules.shouldHaveSize(1)

        val classes = modules[0].classes
        classes.shouldHaveSize(1)

        classes[0].methods.shouldBeEmpty()
    }
}
